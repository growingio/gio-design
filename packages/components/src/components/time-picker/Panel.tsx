import React, { Component } from 'react';
import moment, { Moment } from 'moment';
import classNames from 'classnames';
// import Header from './Header';
import Combobox from './Combobox';
import { PanelProps } from './interface';

function noop() {/* ... */}

function generateOptions(length: number, disabledOptions: number[], hideDisabledOptions: boolean, step = 1) {
  const arr = [];
  for (let value = 0; value < length; value += step) {
    if (!disabledOptions || disabledOptions.indexOf(value) < 0 || !hideDisabledOptions) {
      arr.push(value);
    }
  }
  return arr;
}

function toNearestValidTime(time: Moment, hourOptions: number[], minuteOptions: number[], secondOptions: number[]) {
  const hour = hourOptions.slice().sort((a, b) => Math.abs(time.hour() - a) - Math.abs(time.hour() - b))[0];
  const minute = minuteOptions.slice().sort((a, b) => Math.abs(time.minute() - a) - Math.abs(time.minute() - b))[0];
  const second = secondOptions.slice().sort((a, b) => Math.abs(time.second() - a) - Math.abs(time.second() - b))[0];
  return moment(`${hour}:${minute}:${second}`, 'HH:mm:ss');
}
interface PanelState {
  value: Moment;
  currentSelectPanel: any;
}
class Panel extends Component<PanelProps, PanelState> {
  public static defaultProps = {
    prefixCls: 'rc-time-picker-panel',
    onChange: noop,
    disabledHours: noop,
    disabledMinutes: noop,
    disabledSeconds: noop,
    defaultOpenValue: moment(),
    use12Hours: false,
    addon: noop,
    onKeyDown: noop,
    onAmPmChange: noop,
    inputReadOnly: false,
  };

  public static state = {};

  static getDerivedStateFromProps(props: PanelProps, state: PanelState) {
    if ('value' in props) {
      return {
        ...state,
        value: props.value,
      };
    }
    return null;
  }

  private onChange = (newValue: Moment) => {
    const { onChange } = this.props;
    this.setState({ value: newValue });
    if (onChange) {
      onChange(newValue);
    }
  };

  private onAmPmChange = (ampm: any) => {
    const { onAmPmChange } = this.props;
    onAmPmChange(ampm);
  };

  private onCurrentSelectPanelChange = (currentSelectPanel: any) => {
    this.setState({ currentSelectPanel });
  };

  private disabledHours = () => {
    const { use12Hours, disabledHours } = this.props;
    let disabledOptions = disabledHours && disabledHours();
    if (use12Hours && Array.isArray(disabledOptions)) {
      if (this.isAM()) {
        disabledOptions = disabledOptions.filter((h) => h < 12).map((h) => (h === 0 ? 12 : h));
      } else {
        disabledOptions = disabledOptions.map((h) => (h === 12 ? 12 : h - 12));
      }
    }
    return disabledOptions;
  };

  // https://github.com/ant-design/ant-design/issues/5829
  // private close() {
  //   const { onEsc } = this.props;
  //   onEsc();
  // }

  private isAM() {
    const { defaultOpenValue } = this.props;
    const { value } = this.state;
    const realValue = value || defaultOpenValue;
    return realValue.hour() >= 0 && realValue.hour() < 12;
  }

  public render() {
    const {
      prefixCls,
      className,
      // placeholder,
      disabledMinutes,
      disabledSeconds,
      hideDisabledOptions,
      showHour,
      showMinute,
      showSecond,
      format,
      defaultOpenValue,
      // clearText,
      onEsc,
      addon,
      use12Hours,
      // focusOnOpen,
      // onKeyDown,
      hourStep,
      minuteStep,
      secondStep,
      // inputReadOnly,
      // clearIcon,
    } = this.props;
    const { value } = this.state;
    const disabledHourOptions = this.disabledHours();
    const disabledMinuteOptions = disabledMinutes && disabledMinutes(value ? value.hour() : undefined);
    const disabledSecondOptions =
      disabledSeconds && disabledSeconds(value ? value.hour() : undefined, value ? value.minute() : undefined);
    const hourOptions = generateOptions(24, disabledHourOptions || [], !!hideDisabledOptions, hourStep);
    const minuteOptions = generateOptions(60, disabledMinuteOptions || [], !!hideDisabledOptions, minuteStep);
    const secondOptions = generateOptions(60, disabledSecondOptions || [], !!hideDisabledOptions, secondStep);

    const validDefaultOpenValue =
      defaultOpenValue && toNearestValidTime(defaultOpenValue, hourOptions, minuteOptions, secondOptions);

    return (
      <div className={classNames(className, `${prefixCls}-inner`)}>
        <Combobox
          prefixCls={prefixCls}
          value={value}
          defaultOpenValue={validDefaultOpenValue}
          format={format}
          onChange={this.onChange}
          onAmPmChange={this.onAmPmChange}
          showHour={showHour}
          showMinute={showMinute}
          showSecond={showSecond}
          hourOptions={hourOptions}
          minuteOptions={minuteOptions}
          secondOptions={secondOptions}
          disabledHours={this.disabledHours}
          disabledMinutes={disabledMinutes}
          disabledSeconds={disabledSeconds}
          onCurrentSelectPanelChange={this.onCurrentSelectPanelChange}
          use12Hours={use12Hours}
          onEsc={onEsc}
          isAM={this.isAM()}
        />
        {addon(this)}
      </div>
    );
  }
}

export default Panel;
