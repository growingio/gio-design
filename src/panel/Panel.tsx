import React, { useEffect, useState } from 'react';
import classnames from 'classnames';
import { difference } from 'lodash';
import { useControlledState, usePrefixCls } from '@gio-design/utils';
import Typography from '../typography';
import Tabs, { Tab } from '../tabs';
import Divider from '../divider';
import { PanelProps, TabPaneProps } from './interfaces';
import TabPanel from './TabPanel';
import { Avatar } from '..';

const InnerPanel: React.ForwardRefRenderFunction<HTMLDivElement, PanelProps> = (props, ref) => {
  const {
    title,
    description,
    children,
    footer,
    tabSize = 'normal',
    activeKey,
    defaultActiveKey,
    onTabClick,
    onChange,
    avatar,
    actions,
    bordered = true,
    className,
    style,
  } = props;
  const _prefixCls = usePrefixCls('panel');
  const prefix = (classname?: string) => `${_prefixCls}${classname || ''}`;

  const [currentTabKey, setCurrentTabKey] = useControlledState(activeKey, defaultActiveKey);
  const [currentTabIndex, setCurrentTabIndex] = useState(0);
  const panelChildren = React.Children.toArray(children);

  const panelTabs = panelChildren.filter(
    (child) => React.isValidElement(child) && child.type === TabPanel
  ) as React.ReactElement<TabPaneProps>[];

  const otherChildren = difference(panelChildren, panelTabs);

  const showTabs = panelTabs.length > 1;

  useEffect(() => {
    const _currentIndex = panelTabs.findIndex((tab) => tab.key === `.$${currentTabKey}`);
    if (_currentIndex > -1) {
      setCurrentTabIndex(_currentIndex);
    }
  }, [currentTabKey, panelTabs]);

  const showHeader = title || description || !showTabs;

  const onTabChange = (tabKey: string) => {
    setCurrentTabKey(tabKey);
    onChange?.(tabKey);
  };

  return (
    <div ref={ref} className={classnames(prefix(), className, { [prefix('--bordered')]: bordered })} style={style}>
      <div className={classnames(prefix('__header'), { [prefix('__header--hidden')]: !showHeader })}>
        {avatar && (
          <div className={prefix('__header__avatar')}>
            <Avatar size="large" icon={avatar} mode="square" />
          </div>
        )}
        <div className={prefix('__header__meta')}>
          {title && <div className={classnames(prefix('__header__meta__title'))}>{title}</div>}
          {description && (
            <div className={classnames(prefix('__header__meta__description'))}>
              {typeof description === 'string' ? (
                <Typography.Text lines={3}>{description}</Typography.Text>
              ) : (
                description
              )}
            </div>
          )}
        </div>
        {actions && <div className={classnames(prefix('__header__actions'))}>{actions}</div>}
      </div>
      {showTabs ? (
        <>
          <div className={classnames(prefix('__tabs'))}>
            <Tabs
              value={currentTabKey}
              onChange={onTabChange}
              onClick={onTabClick}
              size={tabSize}
              defaultValue={currentTabKey}
            >
              {panelTabs.map((tab) => {
                const {
                  key: tabKey,
                  props: { name },
                } = tab;
                return (
                  <Tab value={String(tabKey).slice(2)} label={name} key={tabKey} className={prefix('__tabs__tab')} />
                );
              })}
            </Tabs>
            <Divider className={prefix('__tabs__divider')} />
          </div>
          {panelTabs[currentTabIndex]}
        </>
      ) : (
        panelTabs[0]
      )}
      {otherChildren}
      {footer && <div className={prefix('__footer')}>{footer}</div>}
    </div>
  );
};

const Panel = React.forwardRef<unknown, PanelProps>(InnerPanel);

Panel.displayName = 'Panel';

export default Panel;
