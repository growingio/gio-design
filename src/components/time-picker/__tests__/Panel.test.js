import React from 'react';
import { mount, render } from 'enzyme';
import Panel from '../index';

describe('Testing Combobox', () => {
  //   it('should match alert base snapshot.', () => {
  //     const wrapper = render(<Panel />);
  //     expect(wrapper).toMatchSnapshot();
  //   });
  //  props test
  it('props test', () => {
    const onClickMock = jest.fn();
    expect(
      mount(
        <Panel defaultOpenValue={false} use12Hours isAM onAmPmChange={onClickMock} onChange={onClickMock} />
      ).exists('.gio-time-picker-panel-combobox')
    ).toMatchSnapshot();
  });
});
