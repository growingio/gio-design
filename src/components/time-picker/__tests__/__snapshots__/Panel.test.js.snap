// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Combobox props test 1`] = `false`;
